compiler/src/iree/compiler/Codegen/SPIRV/SPIRVLinkExecutables.cpp

```diff
@@ -40,22 +60,110 @@ struct SPIRVLinkExecutablesPass final
     }
     sourceExecutableOps.resize(retainSize);
 
-    // Guess a module name, if needed, to make the output files readable.
-    std::string moduleName = guessModuleName(moduleOp, "spirv_module");
+    // Note that at runtime, for a particular executable, only one variant of it
+    // will be loaded. So, all variants of an executable are expected to provide
+    // the exact same set of entry points; this way we can guarantee no matter
+    // which variant is chosen, we have all entry points to call into. The same
+    // entry point in different variants may have different target requirements
+    // though.
+    //
+    // The input to the linking stage are a collection of executables, each may
+    // have multiple variants, but only ever provide one entry point. Together
+    // with the above restriction, we can link two executables if and only if
+    // their variants have the exact same set of target requirements. Under such
+    // circumstances, we can make sure for a particular target requirement
+    // (loaded as one variant during runtime), we can provide all entry points.
+
+    // Build a map from all variants' target requirements to their wrapping
+    // executable ops.
+    std::map<SmallVector<ExecutableTargetAttr, 0>,
+             SmallVector<IREE::HAL::ExecutableOp>>
+        executableBuckets;
+
+    SmallVector<ExecutableTargetAttr, 0> currentTargets;
+    for (int i = 0, e = sourceExecutableOps.size(); i < e; ++i) {
+      IREE::HAL::ExecutableOp executable = sourceExecutableOps[i];
+
+      // Go through all variants and collect all their target requirements and
+      // sort as the unique key.
```

Comment:
Is this assuming every executable has a variant for every kind of target? I think there was some talk in the past of early deduplication of variants based on what the known requirements are for a particular strategy (e.g. reduction doesn't require cooperative matrix, so don't compile as though you have cooperative matrix).

Reply:
> Is this assuming every executable has a variant for every kind of target?

It depends on what's given from the developer. If the developer only ever specify a particlar target, then our hands are tied and there is no way we can have a variant for each target. If no target is given, then we can be flexible here--my thinking is that we always provide a default that are runnable everywhere, then have some variants using more advanced features.

> I think there was some talk in the past of early deduplication of variants based on what the known requirements are for a particular strategy (e.g. reduction doesn't require cooperative matrix, so don't compile as though you have cooperative matrix).

That's kinda disjoint IMO. You can dedup before compilation, and also after compilation. (Though here we are talking about linking so it's not really dedup; it's more merging exectuables.) I'm not too concerned about duplicated compliation at the moment--think how we are using the compiler. If the developer give no targets, we fan out a list of _minimal_ target environment for a certain pipeline to precisely compile a few variants; there should be no duplication there. If the developer give multiple targets, I'm not expecting them to list like hundreds of targets there. Likely just few. So not a top concern given so many other things to support/optimize first.

Reply:
> It depends on what's given from the developer. If the developer only ever specify a particlar target, then our hands are tied and there is no way we can have a variant for each target. If no target is given, then we can be flexible here--my thinking is that we always provide a default that are runnable everywhere, then have some variants using more advanced features.

Sorry, I wasn't very clear about what I was asking. What I was wondering was whether we could have something like
```
hal.executable {
  #target_1 = hal.executable.variant
  #target_2 = hal.executable.variant
  #default = hal.executable.variant
}
hal.executable {
  #target_1 = hal.executable.variant
  #default = hal.executable.variant
}
```
(i.e. default is supported by both target_2 and target_1, and because there is no option for target_2, it needs to default to the default variant)

Reply:
not sure I follow exactly, but yes @qedawkins each executable may have a different set of variants - if this code is assuming executables have the same variants then we'll need to fix that

(we will prune variants that aren't used and deduplicate variants that don't have any differences, so even if we start with each executable having the same variants it's unlikely we'll end up with that)

Reply:
> What I was wondering was whether we could have something like
> 
> ```
> hal.executable {
>   #target_1 = hal.executable.variant
>   #target_2 = hal.executable.variant
>   #default = hal.executable.variant
> }
> hal.executable {
>   #target_1 = hal.executable.variant
>   #default = hal.executable.variant
> }
> ```
> 
> (i.e. default is supported by both target_2 and target_1, and because there is no option for target_2, it needs to default to the default variant)

I assume `#target_1`/`#target_2`/`#default` are all different. You can certainly have the above inputs; but these two executables won't link together--because if we link them together when you have a runtime environment matching `#target_2`, you'll only have the entry points from the first executable but not the second one. 

The last test added actually shows this.

Reply:
ah ok makes sense, I should have read the tests closer then, sorry about that. I focused too much on code and couldn't see where this was happening.